controller/service.go

```diff
@@ -161,16 +162,44 @@ func (c *controller) convergeBalancer(l log.Logger, key string, svc *v1.Service)
 		lbIngressIPs = append(lbIngressIPs, v1.LoadBalancerIngress{IP: lbIP.String()})
 	}
 	svc.Status.LoadBalancer.Ingress = lbIngressIPs
+	if svc.Annotations == nil {
+		svc.Annotations = make(map[string]string)
+	}
+	svc.Annotations[annotationIPAllocateFromPool] = pool
 	return true
 }
 
 // clearServiceState clears all fields that are actively managed by
 // this controller.
 func (c *controller) clearServiceState(key string, svc *v1.Service) {
 	c.ips.Unassign(key)
+	delete(svc.Annotations, annotationIPAllocateFromPool)
 	svc.Status.LoadBalancer = v1.LoadBalancerStatus{}
 }
 
+// updateService return true if svc's status or annotations changed
+// and update svc. Or nothing to do and return false.
+func (c *controller) updateService(svcRo, svc *v1.Service) bool {
+	isUpdate := false
+	if !reflect.DeepEqual(svcRo.Status, svc.Status) {
+		var st v1.ServiceStatus
+		st, svc = svc.Status, svcRo.DeepCopy()
+		svc.Status = st
+		isUpdate = true
+	}
+
+	if !reflect.DeepEqual(svcRo.Annotations, svc.Annotations) {
+		var annotations map[string]string
+		annotations = svc.Annotations
+		if reflect.DeepEqual(svc, svcRo) {
```

Comment:
is this necessary? We entered here because the annotations are different, so this will always be false

Reply:
Here is a nit, ignore..